crates/runtime/src/datafusion/cluster/physical_plan/optimizer/distribute_file_scan.rs

```diff
@@ -0,0 +1,432 @@
+/// has a `CoalescePartitionsExec` at the top of its plan to signal to the Ballista
+/// distributed planner that it can break the plan into a new stage at that point in time.
+///
+/// The vanilla distributed plan looks like this. This would run as single task on a single node:
+/// ```text
+/// DataSourceExec: file_groups={20 groups ...]}, file_type=parquet
+/// ```
+///
+/// The new distributed plan, that runs as many tasks, across several nodes:
+/// ```text
+/// =========ResolvedStage[stage_id=1.0, partitions=1]=========
+/// ShuffleWriterExec: partitioning:None
+///   EnsureRuntimeDependencyExec: RuntimeDependencySpec {...}
+///     DataSourceExec: file_groups={1 group: [[wiki_a.parquet:0..43660370]]}, file_type=parquet
+///
+/// =========ResolvedStage[stage_id=2.0, partitions=1]=========
+/// ShuffleWriterExec: partitioning:None
+///   EnsureRuntimeDependencyExec: RuntimeDependencySpec {...}
+///     DataSourceExec: file_groups={1 group: [[wiki_a.parquet:43660370..87320740]]}, file_type=parquet
+/// ```
+///
+/// Limits are handled by the default physical pushdown mechanism and are currently
+/// replicated per scan.
+#[derive(Debug)]
+pub struct DistributeFileScanOptimizer {}
+
+impl DistributeFileScanOptimizer {
+    #[must_use]
+    pub fn new() -> Arc<Self> {
+        Arc::new(DistributeFileScanOptimizer {})
+    }
+}
+
+impl DistributeFileScanOptimizer {
+    /// Bytes read is either the whole file or some part of it
+    #[allow(clippy::cast_sign_loss)]
+    fn read_size(pf: &PartitionedFile) -> u64 {
+        if let Some(range) = pf.range.as_ref() {
+            (range.end - range.start) as u64
+        } else {
+            pf.object_meta.size
+        }
+    }
+
+    /// Emit file groups that are constrained by the target byte size
+    fn groups_by_byte_size(
+        files: impl IntoIterator<Item = PartitionedFile>,
+        partition_byte_size: u64,
+    ) -> Vec<FileGroup> {
+        let mut groups: Vec<Vec<PartitionedFile>> = vec![vec![]];
+        let mut current_group_size: u64 = 0;
+        files.into_iter().for_each(|f| {
+            let size = Self::read_size(&f);
+
+            if (current_group_size + size) >= partition_byte_size {
+                groups.push(vec![]);
+                current_group_size = 0;
+            }
+
+            current_group_size += size;
+
+            let Some(last_group) = groups.last_mut() else {
+                unreachable!("There must be at least one group")
+            };
+
+            last_group.push(f);
+        });
+        groups
+            .into_iter()
+            .filter_map(|files| {
+                if files.is_empty() {
+                    None
+                } else {
+                    Some(FileGroup::new(files))
+                }
+            })
+            .collect()
+    }
+
+    /// Group `FileGroup`s into stages based on task level parallelism and optional user-defined
+    /// max stages configuration
+    fn groups_to_stages(
```

Comment:
A `FileGroup` in `DataSourceExec` corresponds to a partition. A collection of FileGroup make up a partitioned file scan. A "stage" here means "independently executable `DataSourceExec`".